File: library/builtins/traits/functions/template.rs
use crate::*;
use wipple::*;

#[derive(TypeInfo, Debug, Clone)]
pub struct Template {
    pub captured_env: Env,
    pub parameters: Vec<(Pattern, String)>,
    pub return_value: Value,
    pub(crate) partially_applied_inputs: Vec<Value>,
}

impl Primitive for Template {}

impl Template {
    fn expand(&self, env: &Env, stack: &Stack) -> Result<Value> {
        assert!(self.parameters.len() == self.partially_applied_inputs.len());

        let inner_env = self.captured_env.child();

        for ((pattern, name), value) in self.parameters.iter().zip(&self.partially_applied_inputs) {
            let value = pattern(value, env, stack)?
                .ok_or_else(|| error("Cannot use this value as input to this template", stack))?
                .into_owned();

            inner_env.set_variable(stack, name, value)?;
        }

        self.return_value
            .interpolate(true, &inner_env, stack)
            .map(Cow::into_owned)
    }
}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    // Template == Function
    env.add_relation_between(stack, |template: Template| {
        Function::new(move |value, env, stack| {
            let mut template = template.clone();
            template.partially_applied_inputs.push(value);

            if template.parameters.len() == template.partially_applied_inputs.len() {
                template
                    .expand(env, stack)?
                    .evaluate(env, stack)
                    .map(Cow::into_owned)
            } else {
                Ok(Value::of(template))
            }
        })
    })?;

    // Template == Text
    env.add_text_relation::<Template>("template", stack)?;

    // TODO: Use 'apply' instead
    env.set_variable(
        stack,
        "expand",
        Value::of(Function::new(move |value, env, stack| {
            let items = value
                .evaluate(env, stack)?
                .get_or::<List>("Expected list", env, stack)?
                .items
                .clone();

            Ok(Value::of(Function::new(move |template, env, stack| {
                let mut template = template
                    .evaluate(env, stack)?
                    .get_or::<Template>("Expected template", env, stack)?
                    .into_owned();

                let expected_count =
                    template.parameters.len() - template.partially_applied_inputs.len();

                if items.len() != expected_count {
                    return Err(error(
                        &format!(
                            "Expected {} parameters, found {}",
                            expected_count,
                            items.len(),
                        ),
                        stack,
                    ));
                }

                template.partially_applied_inputs.append(&mut items.clone());

                template.expand(env, stack)
            })))
        })),
    )?;

    Ok(())
}

File: Library/Sources/Traits/Functions/Template.swift
import Wipple

public struct Template: Primitive {
    public var capturedEnv: Env
    public var parameters: [(pattern: Pattern, name: String)]
    public var returnValue: Value

    internal var partiallyAppliedInputs: [Value] = []

    public init(capturing capturedEnv: Env, parameters: [(pattern: Pattern, name: String)], returning returnValue: Value) {
        self.capturedEnv = capturedEnv
        self.parameters = parameters
        self.returnValue = returnValue
    }
}

private extension Template {
    var isFullyApplied: Bool {
        self.parameters.count == self.partiallyAppliedInputs.count
    }

    func expand(_ env: Env, _ stack: Stack) throws -> Value {
        precondition(self.isFullyApplied)

        let templateEnv = self.capturedEnv.child()

        for (parameter, value) in zip(self.parameters, self.partiallyAppliedInputs) {
            guard let matchedValue = try parameter.pattern.match(value, env, stack) else {
                throw Exit.error("Cannot use this value as input to this template", stack)
            }

            try templateEnv.setVariable(parameter.name, to: matchedValue, stack)
        }

        return try self.returnValue.interpolate(direct: true, templateEnv, stack)
    }
}

internal func setupTemplate(_ env: Env, _ stack: Stack) throws {
    // Template == Function
    try env.addRelation(from: Template.self, to: Function.self, stack) { template in
        Function { value, env, stack in
            var template = template
            template.partiallyAppliedInputs.append(value)

            if template.isFullyApplied {
                return try template.expand(env, stack).evaluate(env, stack)
            } else {
                return Value(template)
            }
        }
    }

    // Template == Text
    try env.addRelation(text: "template", for: Template.self, stack)

    // TODO: Use Apply trait instead
    try env.setVariable(
        "expand",
        to: Value(Function { value, env, stack in
            let items = try value
                .evaluate(env, stack)
                .get(List.self, or: "Expected list", env, stack)
                .rawValue

            return Value(Function { template, env, stack in
                var template = try template
                    .evaluate(env, stack)
                    .get(Template.self, or: "Expected template", env, stack)

                let expectedCount = template.parameters.count - template.partiallyAppliedInputs.count

                guard items.count == expectedCount else {
                    throw Exit.error("Expected \(expectedCount) parameters, found \(items.count)", stack)
                }

                template.partiallyAppliedInputs.append(contentsOf: items)

                return try template.expand(env, stack)
            })
        }),
        stack
    )
}
